File: app-service/src/main/java/com/go/groovy/exception/ServiceException.java
package com.go.groovy.exception;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import java.io.Serial;

@Data
@EqualsAndHashCode(callSuper = true)
@NoArgsConstructor
@AllArgsConstructor
public final class ServiceException extends RuntimeException {

    @Serial
    private static final long serialVersionUID = 1L;

    /**
     * 错误码
     */
    private Integer code;

    /**
     * 错误提示
     */
    private String message;

    public ServiceException(String message) {
        this.message = message;
    }

    public ServiceException(String message, Integer code) {
        this.message = message;
        this.code = code;
    }

    public static void assertFalse(boolean expression, String msg) {
        if (expression) {
            throw new ServiceException(msg);
        }
    }

    public static void throwException(String msg) {
        throw new ServiceException(msg);
    }

    @Override
    public String getMessage() {
        return message;
    }

    public Integer getCode() {
        return code;
    }
}

File: module-common/src/main/java/com/go/common/exception/ServiceException.kt
package com.go.common.exception

import cn.hutool.http.HttpStatus
import java.io.Serializable

/**
 * 服务异常基类
 */
data class ServiceException(
    /**
     * 错误码
     */
    var code: Int = HttpStatus.HTTP_INTERNAL_ERROR,
    /**
     * 错误提示
     */
    var msg: String = "服务异常",
) : RuntimeException(msg), Serializable {

    companion object {
        @JvmStatic
        private val serialVersionUID: Long = 1L

        fun assertTrue(expression: Boolean, msg: String) {
            if (!expression) {
                throw ServiceException(msg = msg)
            }
        }

        fun assertFalse(expression: Boolean, msg: String) {
            if (expression) {
                throw ServiceException(msg = msg)
            }
        }
    }
}